src/main/app-updater.ts

```diff
@@ -1,20 +1,124 @@
-import { autoUpdater } from "electron-updater";
+import { autoUpdater, UpdateInfo } from "electron-updater";
 import logger from "./logger";
+import { NotificationChannelAdd, NotificationChannelPrefix } from "../common/notification-ipc";
+import { ipcMain } from "electron";
+import { isDevelopment, isTestEnv } from "../common/vars";
+import { WindowManager } from "./window-manager";
+import { delay } from "../common/utils";
 
-export class AppUpdater {
-  static readonly defaultUpdateIntervalMs = 1000 * 60 * 60 * 24; // once a day
+class NotificationBackchannel {
+  private static _id = 0;
 
-  static checkForUpdates() {
-    return autoUpdater.checkForUpdatesAndNotify();
+  static nextId(): string {
+    return `${NotificationChannelPrefix}${NotificationBackchannel._id++}`;
   }
+}
+
+const title = "Lens Updater";
+
+async function autoUpdateCheck(windowManager: WindowManager, args: UpdateInfo): Promise<void> {
+  return new Promise(async resolve => {
+    const body = `Version ${args.version} of Lens IDE is now available. Would you like to update?`;
+    const yesNowChannel = NotificationBackchannel.nextId();
+    const yesLaterChannel = NotificationBackchannel.nextId();
+    const noChannel = NotificationBackchannel.nextId();
+
+    function cleanupChannels() {
+      ipcMain.removeAllListeners(yesNowChannel);
+      ipcMain.removeAllListeners(yesLaterChannel);
+      ipcMain.removeAllListeners(noChannel);
+    }
+
+    ipcMain
+      .on(yesNowChannel, async () => {
+        logger.info("[UPDATE CHECKER]: User chose to update immediately");
+        cleanupChannels();
+
+        await autoUpdater.downloadUpdate();
+        autoUpdater.quitAndInstall();
+
+        resolve();
+      })
+      .on(yesLaterChannel, async () => {
+        logger.info("[UPDATE CHECKER]: User chose to update on quit");
+        cleanupChannels();
+
+        await autoUpdater.downloadUpdate();
+        autoUpdater.autoInstallOnAppQuit = true;
 
-  constructor(protected updateInterval = AppUpdater.defaultUpdateIntervalMs) {
-    autoUpdater.logger = logger;
+        resolve();
+      })
+      .on(noChannel, () => {
+        logger.info("[UPDATE CHECKER]: User chose not to update");
+        cleanupChannels();
+        resolve();
+      });
+
+    windowManager.sendToView({
```

Comment:
I feel this is tight coupling of main & renderer & ui. Could we just broadcast a message that an update is available (with update info). Based on that message renderer could show a notification. On notification action renderer could send a message back to main. Meaning: main should not need to know how update information is drawn, it just provides the information and channel for triggering an update process.

Reply:
Do you mean the `status` ad `action` fields below? Because the rest of that data is quite required. I don't think that it is tight coupling give that `main` the source of a notification should decided what is displayed on it.

Though if you are talking about the "sendToView" vs "broadcast". I don't think we should broadcast because then `main` has to worry about setting an ID so that the notification is only displayed once (which I don't think `main` should need to care about)

Reply:
> Do you mean the `status` ad `action` fields below? Because the rest of that data is quite required. I don't think that it is tight coupling give that `main` the source of a notification should decided what is displayed on it.

Not really those specific fields but this logic in general. Why can't `main` just broadcast:
```
broadcastMessage("app:update-available", updateInfo);
```

and a component in renderer could listen this and act accordingly? Why main needs to know about buttons, labels etc.. it should be renderers decision to draw information.


> Though if you are talking about the "sendToView" vs "broadcast". I don't think we should broadcast because then `main` has to worry about setting an ID so that the notification is only displayed once (which I don't think `main` should need to care about)

IMHO main should not care who is listening these kind of messages as long as these messages stay within core application. Even if we send this to just a view how can you be sure that it's displayed only once?

Reply:
> Meaning: main should not need to know how update information is drawn, it just provides the information and channel for triggering an update process.

IMHO this sounds reasonable. Main just send information and renderer decides how the notification looks like. Ideally we should have only one way to trigger those notifications.